src/hotspot/share/classfile/symbolTable.cpp

```diff
@@ -126,38 +126,68 @@ static uintx hash_shared_symbol(const char* s, int len) {
 #endif
 
 class SymbolTableConfig : public AllStatic {
-private:
+
 public:
-  typedef Symbol* Value;  // value of the Node in the hashtable
+  typedef Symbol Value;  // value of the Node in the hashtable
 
   static uintx get_hash(Value const& value, bool* is_dead) {
-    *is_dead = (value->refcount() == 0);
+    *is_dead = (value.refcount() == 0);
     if (*is_dead) {
       return 0;
     } else {
-      return hash_symbol((const char*)value->bytes(), value->utf8_length(), _alt_hash);
+      return hash_symbol((const char*)value.bytes(), value.utf8_length(), _alt_hash);
     }
   }
   // We use default allocation/deallocation but counted
   static void* allocate_node(void* context, size_t size, Value const& value) {
     SymbolTable::item_added();
-    return AllocateHeap(size, mtSymbol);
+    return allocate_node_impl(size, value);
   }
-  static void free_node(void* context, void* memory, Value const& value) {
+  static void free_node(void* context, void* memory, Value & value) {
     // We get here because #1 some threads lost a race to insert a newly created Symbol
     // or #2 we're cleaning up unused symbol.
     // If #1, then the symbol can be either permanent,
     // or regular newly created one (refcount==1)
     // If #2, then the symbol is dead (refcount==0)
-    assert(value->is_permanent() || (value->refcount() == 1) || (value->refcount() == 0),
-           "refcount %d", value->refcount());
-    if (value->refcount() == 1) {
-      value->decrement_refcount();
-      assert(value->refcount() == 0, "expected dead symbol");
+    assert(value.is_permanent() || (value.refcount() == 1) || (value.refcount() == 0),
+           "refcount %d", value.refcount());
+    if (value.refcount() == 1) {
+      value.decrement_refcount();
+      assert(value.refcount() == 0, "expected dead symbol");
+    }
+    if (value.refcount() != PERM_REFCOUNT) {
+      FreeHeap(memory);
+      SymbolTable::item_removed();
+    }
```

Comment:
We don’t call afree() anymore if the symbol is permanent. . Is this intentional?

Actually the old code has a bug. A non permanent symbol allocated from the c heap can overflow the refcount and become permanent. If it needs to be freed, we cannot call afree.

Anyway, how is it possible for a permanent symbol to be freed here? Maybe this should be changed to an assert?

Reply:
Yes, it seemed dubious if we ever freed the entry from the symbol arena, so I think it's ok.  It would only free the Symbol if it was the last one allocated and there was a race when inserting the same symbol permanent symbol in the table.

You're right there would be a bug if we tried to free a CHeap allocated symbol from the arena if the refcount became permanent.  I'm not sure what would happen in that case.

Reply:
Oh but if it overflows the refcount, we won't free it, so it's not a bug.

Reply:
From the comment above this code, it seems we do free permanent symbols, but only when the current thread lost when competing to insert a symbol for the boot loader. So the old code is correct, and the new code would have a leak.

Reply:
I've added back the `Afree()` for permanent symbols. It passed tier1 testing. Running other tiers testing currently.